### PR TITLE
address: use resolveProtoAddress to resolve listener addresses

### PR DESCRIPTION
**Pull request by hennna**
Signed-off-by: Henna Huang <henna@google.com>

*title*: *Use resolveProtoAddress to resolve listener addresses*

*Description*:
In https://github.com/envoyproxy/envoy/pull/1835, support for synchronous address resolvers and upstream address resolution were added. This PR extends support to resolving listener addresses.

*Risk Level*: Low 

*Testing*:
Existing tests pass.
listener_impl test cases added.

*Docs Changes*:
n/a

*Release Notes*:
n/a

### CODE REVIEW

File: test/server/listener_manager_impl_test.cc [hunk 1]
```diff
@@ -1512,6 +1512,23 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, FreebindListenerEnabled) {
...
+    name: AddressResolverdListener
+    address:
+      socket_address: { address: 127.0.0.1, port_value: 1111, resolver_name: envoy.ip }
+    filter_chains:
+    - filters:
+  )EOF",
+                                                       Network::Address::IpVersion::v4);
+
+  EXPECT_CALL(listener_factory_, createListenSocket(_, _, true));
+  manager_->addOrUpdateListener(parseListenerFromV2Yaml(yaml), true);
```
Comment (hennna) [inline test/server/listener_manager_impl_test.cc]: This doesn't test that the resolved address matches what is expected. Is that necessary?

File: test/server/listener_manager_impl_test.cc [hunk 2]
```diff
@@ -123,6 +127,13 @@ class ListenerManagerImplWithRealFiltersTest : public ListenerManagerImplTest {
   }
 };
 
+class MockAddressResolver : public Network::Address::Resolver {
```
Comment (mattklein123) [inline test/server/listener_manager_impl_test.cc]: test nit: I would move this into the common mocks as it is generally useful

File: test/server/listener_manager_impl_test.cc [hunk 3]
```diff
@@ -1512,6 +1523,31 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, FreebindListenerEnabled) {
...
+    name: AddressResolverdListener
+    address:
+      socket_address: { address: 127.0.0.1, port_value: 1111, resolver_name: envoy.mock.resolver }
+    filter_chains:
+    - filters:
+  )EOF",
+                                                       Network::Address::IpVersion::v4);
+
+  MockAddressResolver mock_resolver;
+  EXPECT_CALL(mock_resolver, name()).WillRepeatedly(Return("envoy.mock.resolver"));
```
Comment (mattklein123) [inline test/server/listener_manager_impl_test.cc]: I would probably just setup a default ON_CALL in the mock for this one. (EXPECT_CALL/WillRepeatedly is a bit of an anti-pattern)

File: test/server/listener_manager_impl_test.cc (hunk 2 above)
Comment (hennna) [reply test/server/listener_manager_impl_test.cc]: thanks for the review. done.

**Pull request merged**